---
title: "Introduction_to_KSAPA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_to_KSAPA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# Introduction

Long read RNA sequencing is a powerful technology that can be used to examine gene expression level as well post-transcriptional processing at full length of single RNA molecule with much less ambiguity when compared with short read RNA sequencing. Among various long read sequencing strategy, direct RNA sequencing usually provide the more accurate 3' end information. 

Bioinformatic pipelines specifically designed for long read RNA seq data ultilising the full length RNA information are very limited. The KSAPA package provides a novel quantification pipeline to extract polyA site usage information from long read RNAseq data sets. KSAPA searches for 3'end peaks (polyA site, PAS) of sequence reads  across the whole data set and then quantifies with the usage of the peaks at single gene level. 


# KSAPA Installation

KSAPA can be installed from github 

```
install.packages("devtools")
devtools::install_github("VCCRI/Sierra", build = TRUE, build_vignettes = TRUE, build_opts = c("--no-resave-data", "--no-manual"))

library(Sierra)
```


```
install.packages("devtools")
devtools::install_github("VCCRI/Sierra", build = TRUE, build_vignettes = TRUE
```

```{r setup}
library(KSAPA)
```
# Preparing input data sets

For this vignette, we will be using cardiac scRNA-seq from [Farbehi et al.](https://elifesciences.org/articles/43882). This study profiled cardiac interstitial cells from the hearts of mice subject to a myocardial infarction (MI) injury or sham control surgery. For the purpose of this vignette we have provided a small subset of the data from the sham condition and one of the MI time-points.    

## Required inputs


### BAM file
BAM files should be created from CellRanger or an equivalent program. Sierra requires BAM files to include two meta-data field tags, unique molecular identifier (UMI tag ``UB:Z``) and cell barcode (tag ``CB:Z``), which enables cell type identification and peak quantification. If your BAM file utilizes different field tags, please see section [below](#BAM_field). 

### Splice junctions file
A BED formatted file that represents splice junctions that occur within the BAM file. This file can be created using the regtools program which needs to be run from your system command line. We specify the strand of the reads using the option ``-s RF`` (first-strand/RF). Note that on older versions of retools this is specified with ``-s 1``. Below is an example of command line syntax that will generate the required output file. Note for the purposes of this vignette this step can be skipped as both output files, ``Vignette_example_TIP_sham_junctions.bed`` and ``Vignette_example_TIP_MI_junctions.bed``, have already been precomputed.
