#' Visualize the 3'end APA changes between the two samples
#' 
#' This function loads a gtf file and extract necessary information of each gene from it.
#' @param APA_data APA data generated by APA_profile function
#' @param P_cutoff adjusted P value cutoff to be considered as significant change
#' @param delta PAU change cutoff between two samples to be considered as significant APA events
#' @param FC fold change cutoff of PAU between two samples to be considered as significant APA events
#' @param position either distal or proximal PAS is to be examined
#' @param type either FC or delta is used to call significant APA events
#' @param cores number of threads used for the computation
#' @return a data table showing the significant APA change between group1 and group2
#' @export

end_PAS_examine <- function(PAU_data,reads,P_cutoff=0.05,FC=1.5,delta=10,control,experimental, position="distal", type="delta",cores=1){
  samples1 <- unique(reads[treatment==control]$sample)
  samples2 <- unique(reads[treatment==experimental]$sample)
  samples1_col <- paste0(samples1, " PAU", sep="")
  samples2_col <- paste0(samples2, " PAU", sep="")
  genes <- unique(PAU_data$gene_id)
  dPAU_data <-data.table(matrix(-100, ncol = length(samples1_col)+length(samples2_col), nrow = length(genes)))
  colnames(dPAU_data) <- c(samples1_col,samples2_col)
  dPAU_data$gene_id <- genes
  PAU_test_fun<-function(gene){
    if (position=="distal"){
      print(gene)
      target_table <-PAU_data[gene_id==gene]
      if(nrow(target_table)>1){
        if(all(target_table$strand=="+")){
          distal <-max(target_table$PAS)
        }else{
          distal <-min(target_table$PAS)
        }
        gene_data <- as.numeric(target_table[PAS == distal,.SD,.SDcols = c(samples1_col,samples2_col)])
        test <- t.test(gene_data[1:length(samples1)], 
                       gene_data[(length(samples1_col)+1):(length(samples1_col)+length(samples2_col))])
        target_table <- target_table[PAS==distal]
        target_table[gene_id==gene, c(paste0("mean of ",control),paste0("mean of ",experimental),"pvalue") := 
                       list(as.numeric(test$estimate[1]),as.numeric(test$estimate[2]), as.numeric(test$p.value))]
        return(target_table)
      }
    } 
    if (position=="proximal") {
      print(gene)
      target_table <-PAU_data[gene_id==gene]
      if(nrow(target_table)>1){
        if(all(target_table$strand=="+")){
          distal <-min(target_table$PAS)
        }else{
          distal <-max(target_table$PAS)
        }
        gene_data <- as.numeric(target_table[PAS == distal,.SD,.SDcols = c(samples1_col,samples2_col)])
        test <- t.test(gene_data[1:length(samples1)], 
                       gene_data[(length(samples1_col)+1):(length(samples1_col)+length(samples2_col))])
        target_table <- target_table[PAS==distal]
        target_table[gene_id==gene, c(paste0("mean of ",control),paste0("mean of ",experimental),"pvalue") := 
                       list(as.numeric(test$estimate[1]),as.numeric(test$estimate[2]), as.numeric(test$p.value))]
        return(target_table)
      }
    }
  }
  output <- pbmclapply(genes, PAU_test_fun, mc.cores = cores)
  end_PAU_df <- rbindlist(output, fill = T)
 
  APA_table <-end_PAU_df
  APA_table$P_adj <-p.adjust(APA_table$pvalue, method = "fdr")
  APA_table$Fold_change<- APA_table[,.SD,.SDcols=c(paste0("mean of ",experimental))]/APA_table[,.SD,.SDcols=c(paste0("mean of ",control))]
  APA_table$Vol_P <- (-log10(APA_table$P_adj))
  APA_table[Vol_P == Inf, Vol_P := (max(-log10(APA_table[P_adj!=0]$P_adj))+0.1)]
  APA_table$PAU_change <- (APA_table[,.SD,.SDcols=c(paste0("mean of ",experimental))]-APA_table[,.SD,.SDcols=c(paste0("mean of ",control))])
  
  if(type=="FC"){
    APA_table$Col <- "gray"
    APA_table$APA_trend <- "no change"
    APA_table[(P_adj < P_cutoff & Fold_change > FC), Col := "red"]
    APA_table[(P_adj < P_cutoff & Fold_change > FC), APA_trend := "longer"]
    APA_table[(P_adj < P_cutoff & Fold_change < 1/FC), Col := "blue"]
    APA_table[(P_adj < P_cutoff & Fold_change < 1/FC), APA_trend := "shorter"]
    shortening <- length(subset(APA_table,Col=="blue")$gene_id)
    lengthening <- length(subset(APA_table,Col=="red")$gene_id)
    all <- length(APA_table$gene_id)
    names <-grep("^PAUs_", colnames(APA_table), value = TRUE)
    
    #  pdf("APA_plot.pdf",7,7)
    par(mar = c(6, 5, 4, 3) + 0.1)
    plot(x=log2(APA_table$Fold_change),  y=APA_table$Vol_P,
         main =paste0("APA trend (",control, " vs ", experimental, ")", sep=""),  
         pch= 20, sub = paste(shortening,"shortened","and",lengthening,"lengthend" ,"in all", all, "events",sep=" "),
         xlab = paste0(position, " PAU Fold change (",experimental, "/", control, ")", sep=""),  ylab = "-log10(adjusted p value)",  col = APA_table$Col, cex=0.7)
    abline(v = c(-log2(FC),log2(FC)), col = "green", lty = 2, lwd = 2)
    abline(h=(-log10(P_cutoff)), col = "green", lty = 2, lwd = 2)
    #  dev.off()
  }
  
  if(type=="delta"){
    APA_table$Col <- "gray"
    APA_table$APA_trend <- "no change"
    APA_table[(P_adj < P_cutoff & PAU_change > delta), Col := "red"]
    APA_table[(P_adj < P_cutoff & PAU_change > delta), APA_trend := "longer"]
    APA_table[(P_adj < P_cutoff & PAU_change < (-delta)), Col := "blue"]
    APA_table[(P_adj < P_cutoff & PAU_change < (-delta)), APA_trend := "shorter"]
    shortening <- length(subset(APA_table,Col=="blue")$gene_id)
    lengthening <- length(subset(APA_table,Col=="red")$gene_id)
    all <- length(APA_table$gene_id)
    names <-grep("^PAUs_", colnames(APA_table), value = TRUE)
    
    #  pdf("APA_plot.pdf",7,7)
    par(mar = c(6, 5, 4, 3) + 0.1)
    plot(x=APA_table$PAU_change,  y=APA_table$Vol_P,
         main =paste0("APA trend (",control, " vs ", experimental, ")", sep=""),  
         pch= 20, sub = paste(shortening,"decreased","and",lengthening,"increased" ,"in all", all, "events",sep=" "),
         xlab = paste0(position, " PAU change (",experimental, "-", control, ")", sep=""),  ylab = "-log10(adjusted p value)",  col = APA_table$Col, cex=0.7)
    abline(v = c(-delta,delta), col = "green", lty = 2, lwd = 2)
    abline(h=(-log10(P_cutoff)), col = "green", lty = 2, lwd = 2)
    #  dev.off()
  }
  
  return(APA_table[, !c("Col","Vol_P"), with = FALSE])
}